gateway-js/src/index.ts

```diff
@@ -449,13 +502,41 @@ export class ApolloGateway implements GraphQLService {
     }
   }
 
+  private async loadManuallyManaged(config: ManuallyManagedSupergraphSdlGatewayConfig) {
+    try {
+      const result = await config.supergraphSdl({
+        update: this.updateWithSupergraphSdl.bind(this),
```

Comment:
Highest level note: let's make sure that the following ways of calling `update` lead to immediate thrown errors:
- calling it before `config.supergraphSdl` resolves
- calling it concurrently with itself (eg, while `updateWithSupergraphSdl` is running health checks)
- calling it re-entrantly from itself (eg, from the onSchemaChange callbacks or whatever)

This probably involves adding a new "update in progress" `phase` to the state. It's possible that we could *replace* the polling/waiting to poll phases with this phase (and `loaded`) and reimplement `pollServices` as an async `while` loop that calls an `update` function and differentiates between "waiting to poll" and "polling" just by the position in the loop?

Reply:
Ok so I took some steps toward this which might have eliminated the need to worry about concurrency (and I think re-entrant?...need to spend a little more time thinking about it).

Anyway - I think it made a lot of sense to factor out the `await`ed health check bit to make the `updateXyz` functions sync, but with effectively no change in behavior. I think this simplifed some bits nicely and separated the concern of updating the schema from performing health checks. The end goal (which is now the case) is that the `update` callback we hand off to users is also sync!

I now provide a `healthCheck` async function to the user as well, since there's no other path to the existing health check functionality via manual updates as they are now. I think the only thing I need to add is some warning for when users provide both a `supergraphSdl` fn AND specify `config.serviceHealthCheck` since that combination won't do anything.

c64be0f